Add fast sync description. (#26)
subspace/protocol-specs: docs/consensus/consensus_chain.md

4. DSN sync must be able to terminate early if local chain already contains imported blocks that DSN sync was about to download (doesn’t happen often, but possible)
5. Node blocks that are finalized and pruned must be much higher than archiving point such that block available through DSN sync and regular Substrate sync have significant overall (5 archived segments worth of blocks right now)

## Fast sync

> 1. Obtain segment headers from DSN: 1-4 from `sync from DSN implementation`

Reviewer: "sync from DSN implementation" here should be a link, not code block, right?:
```suggestion
1. Obtain segment headers from DSN (steps 1-4 from [Sync from DSN implementation](#sync-from-dsn-implementation))`
```

And similarly about code blocks below.

Author: Sure

2. If the last segment index is greater than 2 - download the last two segments from DSN otherwise exit `Fast sync` and continue with either `sync from DSN implementation` or `Substrate sync`.

> 3. Reconstruct blocks from the two downloaded segments.

Reviewer: Can you remind me why do we need two segments instead of just latest one? I thought padding block with zeroes removed the need to download an extra segment.

Author: We need to "force import" one block and the whole segment after that to keep the archiver valid. Potentially, we can get a single block by a special block request (not publicly exposed in Substrate yet) but I'm not sure the "padding optimization" alone would be enough.

Reviewer: Why wouldn't it be? We literally throw away extra bytes at the beginning anyway: https://github.com/subspace/subspace/blob/2d1e4dc6a44d088c774aee8735c5e035948c8023/crates/subspace-archiving/src/archiver.rs#L329

Author: We likely could use the padding mechanism you described however we need the block from the previous segment to "force import".  And since we have the block already it makes sense to use it as it is.

Reviewer: I thought about it some more and even following that logic why would we download two segments if we can download just one and then import just last full block of the last segment and switch to regular sync from there? Switching to sync from DSN also doesn't make a lot of sense since we already use latest segment, we can switch to Substrate sync immediately.

Author: In theory, we need three "special" blocks - one for "forced import", a block to instantiate an archiver from the correct point (first of the segment), and a block for "NetworkBroadcast". Maybe we can reuse either the first two or the last two but we need an additional block.

> Switching to sync from DSN also doesn't make a lot of sense since we already use the latest segment, we can switch to Substrate sync immediately.

Also, after we downloaded a segment there could be a new segment or (segments) already.

Reviewer: > In theory, we need three "special" blocks - one for "forced import", a block to instantiate an archiver from the correct point (first of the segment), and a block for "NetworkBroadcast". Maybe we can reuse either the first two or the last two but we need an additional block.

Theoretically we just need exacty one: to force import the first block of the beginning of the new segment. Everything else after that will "just work". ver already knows how to initialize from such state, we just need to teach it to "re-initialize" itself if it had block N and then suddenly got block N+100500 and some blocks in between are missing, while corresponding segment headers are already stored.

> Also, after we downloaded a segment there could be a new segment or (segments) already.

Full nodes intentionally store a few segments worth of blocks locally, even in somewhat rare case when this happens (due to how little we need to download) this will not be an issue in practice.

## Fast sync

> 1. Obtain segment headers from DSN: 1-4 from [sync from DSN implementation](#Synchronization)

Reviewer: ```suggestion
1. Obtain segment headers from DSN as described in steps 1 to 4 of [sync from DSN implementation](#Synchronization)
```

> 2. If the last segment index is greater than 2 - download the last two segments from DSN otherwise exit [Fast sync](#fast-sync) and continue with either [sync from DSN implementation](#Synchronization) or [Substrate Sync](#substrate-sync) . Reconstruct blocks from the two downloaded segments. Note: we need only a single last block from the second last segment and we can optimize the solution if acquire this block from Substrate network instead.

Reviewer: This is a better version, but I meant slightly differently:
```suggestion
2. Download and reconstruct all blocks from the last segment of archived history
    * Note: In most cases it'll be necessary to download second last segment as well due to the first block being partially included in latest segment
```

The idea is that we should convey what we are trying to achieve. And we want all blocks from the last segment. The fact that segment before that is also needed is a technical implementation detail.

> 3. Import the last block from the second last segment into the blockchain DB bypassing the blockchain checks of missing parent block. Import the state using [Substrate Sync](#substrate-sync) for the newly imported block. Note: block and its state should be imported as atomic operation.
> 4. Import and execute all blocks from the last segment in regular mode.
> 5. Pass the control to [sync from DSN implementation](#Synchronization). It will either download the new archived segment if any or pass the control to [Substrate Sync](#substrate-sync).

Reviewer: This repeats previous item somewhat and can be made more conscise, also when downloading state we are talking about state sync specifically, not what "Substrate Sync" section below describes:
```suggestion
3. Download state that corresponds to the first block received in the previous step using Substrate State Sync
4. Import the first block of the last segment with its state into the blockchain DB bypassing the blockchain checks of missing parent block, it is important for this to be an atomic operation
5. Import and execute other remaining blocks from the last segment as they would normally
6. Pass the control to [sync from DSN implementation](#Synchronization). It will either download the new archived segment if any or pass the control to [Substrate Sync](#substrate-sync).
```
